discovery/reliable_sender.go

```diff
@@ -0,0 +1,281 @@
+package discovery
+
+import (
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/lnpeer"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// reliableSenderCfg contains all of necessary items for the reliableSender to
+// carry out its duties.
+type reliableSenderCfg struct {
+	// NotifyWhenOnline is a function that allows the gossiper to be
+	// notified when a certain peer comes online, allowing it to
+	// retry sending a peer message.
+	//
+	// NOTE: The peerChan channel must be buffered.
+	//
+	// TODO(wilmer): use [33]byte to avoid unnecessary serializations.
+	NotifyWhenOnline func(peer *btcec.PublicKey, peerChan chan<- lnpeer.Peer)
+
+	// NotifyWhenOffline is a function that allows the gossiper to be
+	// notified when a certain peer disconnects, allowing it to request a
+	// notification for when it reconnects.
+	NotifyWhenOffline func(peerPubKey [33]byte) <-chan struct{}
+
+	// MessageStore is a persistent storage of gossip messages which we will
+	// use to determine which messages need to be resent for a given peer.
+	MessageStore GossipMessageStore
+
+	// IsMsgStale determines whether a message retrieved from the backing
+	// MessageStore is seen as stale by the current graph.
+	IsMsgStale func(lnwire.Message) bool
+
+	// Quit is a channel we'll use to signal that we should tear down all
+	// active peerHandlers.
+	Quit chan struct{}
+
+	// WaitGroup is used to ensure we wait until all active peerHandlers
+	// are completely torn down.
+	WaitGroup sync.WaitGroup
```

Comment:
Could be made its own internal quit and wg, as in other subsystems.

Reply:
Is there a reason for not wanting to keep it as is? We'd have to introduce a `Start` method that doesn't do anything, and a `Stop` method that only closes the `quit` chan.

Reply:
i think it's safer to give it it's own quit and waitgroup, as it's easier to reason about where the calls to `Add` and `Done` are. fwiw, this probably won't behave as expected, we'd need a reference to `sync.WaitGroup`. rn this will _copy_ the waitgroup's state, so the waitgroup in the gossiper will actually be distinct from the waitgroup in the reliable sender

Reply:
> A WaitGroup must not be copied after first use.

From the sourcegraph tool-tip :)

Reply:
Gotcha, will fix.